src/odemis/acq/stitching/_registrar.py

```diff
@@ -14,18 +14,442 @@
+                self.coords_hor[row][col] = x, y
+       
+            # calculate the new position based on the shift with respect to the left neighbor position (prev_pos)
+            pos = (int(prev_pos[0] + (self.size * (1-self.ovrlp)) - x), int(prev_pos[1] - y))
+        
+        return pos, match
+
+    def _pos_to_top(self, row, col):
+        """
+        Returns the position of the new tile after calculating the shift with the
+        neighbor on the top
+        """
+        # shift compared to top
+        x, y = 0, 0
+        
+        # calculate the shift compared to the neighbor on the top if possible
+        pos = None
+        match = 0
+
+        tile = self.tiles[row][col]
+        # the expected shift vector between the two tiles if the scanning would be ideal
+        exp_shift = (0,self.tsize)
+        if row > 0:
+            prev_pos = self.shifts[row - 1][col]
+            prev_tile = self.tiles[row - 1][col]
+            x, y = self._get_shift(prev_tile, tile, exp_shift)
+            match = self._estimateMatch(prev_tile, tile, (exp_shift[0] - x, exp_shift[1] - y))
+            
+            if match < BEST_MATCH and (row > 1 or col > 1):
+                # if the shift calculated does not give a sufficient match,
+                # try the vertical mean shift along this row and column if possible
+                try:
+                    mean_x, mean_y = self._mean_shift(row, col, self.coords_ver)
+                except ValueError:
+                    # if a mean value is not available try the last vertical shift measured
+                    mean_x, mean_y = self.last_ver
+                match_c = self._estimateMatch(prev_tile, tile, (exp_shift[0] - mean_x, exp_shift[1] - mean_y))
+                if match_c > match:
+                    x, y = mean_x, mean_y
+                    match = match_c
+
+                if (mean_x, mean_y) != (0, 0):
+                    # try to calculate shift from here, maybe it works better now
+                    drift = self._get_shift(prev_tile, tile, (exp_shift[0] - mean_x, exp_shift[1] - mean_y))
+                    mean_x += drift[0]
+                    mean_y += drift[1]
+                    match_c = self._estimateMatch(prev_tile, tile, (exp_shift[0] - mean_x, exp_shift[1] - mean_y))
+                    if match_c > match:
+                        x, y = mean_x, mean_y
+                        match = match_c
+            
+            # store the new vertical shift computed
+            self.last_ver = x, y
+            if match >= BEST_MATCH:
+                self.coords_ver[row][col] = x, y
+
+            # calculate the new position based on the shift with respect to the top neighbor position (prev_pos)
+            pos = (int(prev_pos[0] - x), int(prev_pos[1] + (self.size * (1-self.ovrlp)) - y))
+
+        return pos, match
+
+    def _stitch(self, tile, row, col):
+        """
+        Stitches one tile to the overall image. Tiles should be inserted in an
```

Comment:
Is the requirement that the tiles are inserted so that the top and left images are always available still true? If so, it should be clearly stated in addTile() too.... but that's not going to be very handy if we want to support "back & forth" scanning. Alternatively, the method could be either made more clever, or the call to this method delayed until this requirement is fulfilled.